File: legacy/python/src/extractors/video_extractor.py
"""Main video extraction pipeline"""
from typing import List, Dict, Optional
from datetime import datetime
from rich.console import Console

from ..api.youtube_client import YouTubeClient
from ..extractors.transcript_extractor import TranscriptExtractor
from ..extractors.whisper_extractor import WhisperTranscriptExtractor
from ..parsers.llm_parser import GeminiParser
from ..parsers.description_parser import DescriptionParser
from ..database.connection import DatabaseManager
from ..database.repository import (
    VideoRepository,
    StatisticsRepository,
    TranscriptRepository,
    EntityRepository,
    TagRepository,
    PlaylistRepository,
    PlaylistItemRepository,
    ExtractionJobRepository,
    AIAnalysisRepository
)

console = Console()


class VideoExtractor:
    """Main extraction pipeline for videos"""
    
    def __init__(
        self,
        youtube_client: YouTubeClient,
        db_manager: DatabaseManager,
        auto_transcript: bool = True,
        auto_llm_parse: bool = True,
        gemini_api_key: str = None,
        languages: List[str] = None,
        transcript_rate_limit: float = 2.0,
        config: Dict = None
    ):
        """
        Initialize video extractor
        
        Args:
            youtube_client: Authenticated YouTube API client
            db_manager: Database manager instance
            auto_transcript: Automatically extract transcripts
            auto_llm_parse: Automatically parse transcripts with LLM
            gemini_api_key: Gemini API key for LLM parsing
            languages: Preferred transcript languages
            transcript_rate_limit: Delay between transcript requests (seconds, default: 2.0)
            config: Full configuration dictionary (for Whisper settings)
        """
        self.youtube_client = youtube_client
        self.db_manager = db_manager
        self.auto_transcript = auto_transcript
        self.auto_llm_parse = auto_llm_parse
        
        # Initialize Whisper if enabled
        whisper_extractor = None
        if config and config.get('extraction', {}).get('whisper', {}).get('enabled', False):
            try:
                whisper_extractor = WhisperTranscriptExtractor(config)
                console.print("[dim]Whisper fallback enabled[/dim]")
            except Exception as e:
                console.print(f"[yellow]Warning: Could not initialize Whisper: {e}[/yellow]")
        
        self.transcript_extractor = TranscriptExtractor(
            languages, 
            rate_limit_delay=transcript_rate_limit,
            whisper_extractor=whisper_extractor
        )
        self.description_parser = DescriptionParser()
        
        if auto_llm_parse:
            try:
                # Get model name from config
                model_name = config.get('api', {}).get('gemini_model', 'gemini-3-flash-preview') if config else 'gemini-3-flash-preview'
                self.llm_parser = GeminiParser(api_key=gemini_api_key, model=model_name)
            except ValueError as e:
                console.print(f"[yellow]Warning: {e}[/yellow]")
                console.print("[yellow]LLM parsing will be disabled[/yellow]")
                self.auto_llm_parse = False
                self.llm_parser = None
        else:
            self.llm_parser = None
    
    def extract_single_video(
        self,
        video_id: str,
        skip_transcript: bool = False,
        skip_llm: bool = False
    ) -> Optional[Dict]:
        """
        Extract complete data for a single video
        
        Args:
            video_id: YouTube video ID
            skip_transcript: Skip transcript extraction
            skip_llm: Skip LLM parsing
            
        Returns:
            Dictionary with extraction results or None if failed
        """
        console.print(f"\n[bold blue]Extracting video: {video_id}[/bold blue]")
        
        # Step 1: Get video metadata
        console.print("  [cyan]Fetching metadata...[/cyan]")
        video_data = self.youtube_client.get_single_video(video_id)
        
        if not video_data:
            console.print(f"  [red]Failed to fetch video metadata[/red]")
            return None
        
        # Sanitize output for Windows PowerShell
        title = video_data['title'][:60].encode('ascii', 'replace').decode('ascii')
        channel = video_data['channel_title'].encode('ascii', 'replace').decode('ascii')
        console.print(f"  [green][OK][/green] Title: {title}")
        console.print(f"  [green][OK][/green] Channel: {channel}")
        console.print(f"  [green][OK][/green] Duration: {video_data['duration_seconds']}s")
        
        # Save to database
        with self.db_manager.get_session() as session:
            # Create or update video
            video = VideoRepository.create_or_update(session, {
                'video_id': video_data['video_id'],
                'title': video_data['title'],
                'description': video_data['description'],
                'channel_id': video_data['channel_id'],
                'channel_title': video_data['channel_title'],
                'published_at': datetime.fromisoformat(video_data['published_at'].replace('Z', '+00:00')),
                'duration': video_data['duration'],
                'duration_seconds': video_data['duration_seconds'],
                'is_short': video_data['is_short'],
                'category_id': video_data['category_id'],
                'definition': video_data['definition'],
                'caption': video_data['caption'],
                'licensed_content': video_data['licensed_content']
            })
            
            # Add statistics snapshot
            StatisticsRepository.add_snapshot(session, video_id, {
                'view_count': video_data['view_count'],
                'like_count': video_data['like_count'],
                'comment_count': video_data['comment_count']
            })
            
            # Add original tags
            if video_data.get('tags'):
                TagRepository.add_tags_to_video(session, video_id, video_data['tags'])
        
        # Step 2: Parse description for GitHub repos and URLs (always runs)
        console.print("  [cyan]Parsing description...[/cyan]")
        description_parsed = self.description_parser.parse(
            video_data['title'],
            video_data.get('description', '')
        )
        
        if description_parsed['github_repos'] or description_parsed['websites']:
            console.print(f"  [green][OK][/green] Found {len(description_parsed['github_repos'])} GitHub repos")
            console.print(f"  [green][OK][/green] Found {len(description_parsed['websites'])} websites")
            
            with self.db_manager.get_session() as session:
                # Add entities from description
                desc_entities = self.description_parser.extract_entities_for_database(description_parsed)
                if desc_entities:
                    EntityRepository.add_entities(session, video_id, desc_entities)
                
                # Add tags
                desc_tags = self.description_parser.get_tags(description_parsed)
                if desc_tags:
                    TagRepository.add_tags_to_video(session, video_id, desc_tags)
        else:
            console.print("  [dim][SKIP][/dim] No links found in description")
        
        # Step 3: Extract transcript
        transcript_data = None
        if self.auto_transcript and not skip_transcript:
            console.print("  [cyan]Extracting transcript...[/cyan]")
            transcript_data = self.transcript_extractor.extract(video_id)
            
            if transcript_data:
                source = "Whisper" if transcript_data.get('from_whisper') else "YouTube"
                console.print(f"  [green][OK][/green] Transcript extracted via {source} ({len(transcript_data['full_text'])} chars)")
                
                with self.db_manager.get_session() as session:
                    TranscriptRepository.create(session, video_id, transcript_data)
            else:
                console.print("  [yellow][WARN][/yellow] No transcript available")
        
        # Step 4: Parse with LLM
        parsed_entities = None
        if self.auto_llm_parse and not skip_llm and transcript_data:
            console.print("  [cyan]Parsing with Gemini AI...[/cyan]")
            
            try:
                parsed_entities = self.llm_parser.parse_transcript(
                    transcript_data['full_text'],
                    video_data['title']
                )
                
                console.print(f"  [green][OK][/green] Found {len(parsed_entities['topics'])} topics")
                console.print(f"  [green][OK][/green] Found {len(parsed_entities['github_repos'])} repos")
                console.print(f"  [green][OK][/green] Found {len(parsed_entities['websites'])} websites")
                console.print(f"  [green][OK][/green] Found {len(parsed_entities['people'])} people")
                
                with self.db_manager.get_session() as session:
                    # Delete old entities
                    EntityRepository.delete_by_video(session, video_id)
                    
                    # Add new entities
                    entities = self.llm_parser.extract_entities_for_database(parsed_entities)
                    EntityRepository.add_entities(session, video_id, entities)
                    
                    # Add tags
                    tags = self.llm_parser.get_tags(parsed_entities)
                    if tags:
                        TagRepository.add_tags_to_video(session, video_id, tags)
                    
                    # Add AI analysis
                    analysis = self.llm_parser.get_analysis(parsed_entities)
                    AIAnalysisRepository.create_or_update(session, video_id, analysis)
            
            except Exception as e:
                # Error details already shown by llm_parser
                console.print(f"  [yellow]Skipping AI analysis[/yellow]")
        
        console.print(f"  [bold green][OK] Extraction complete![/bold green]\n")
        
        return {
            'video_data': video_data,
            'transcript_data': transcript_data,
            'parsed_entities': parsed_entities
        }
    
    def extract_playlist(
        self,
        playlist_id: str,
        skip_existing: bool = True,
        max_videos: int = None
    ) -> Dict:
        """
        Extract all videos from a playlist
        
        Args:
            playlist_id: YouTube playlist ID
            skip_existing: Skip videos already in database
            max_videos: Maximum number of videos to process
            
        Returns:
            Dictionary with extraction summary
        """
        console.print(f"\n[bold blue]Extracting playlist: {playlist_id}[/bold blue]\n")
        
        # Get playlist info
        playlist_info = self.youtube_client.get_playlist_info(playlist_id)
        
        if not playlist_info:
            console.print("[red]Failed to fetch playlist info[/red]")
            return {'success': False, 'error': 'Playlist not found'}
        
        # Sanitize playlist title for Windows PowerShell
        playlist_title = playlist_info['title'].encode('ascii', 'replace').decode('ascii')
        console.print(f"[bold]{playlist_title}[/bold]")
        console.print(f"Videos: {playlist_info['video_count']}\n")
        
        # Save playlist to database
        with self.db_manager.get_session() as session:
            PlaylistRepository.create_or_update(session, {
                'playlist_id': playlist_id,
                'title': playlist_info['title'],
                'description': playlist_info.get('description', ''),
                'video_count': playlist_info['video_count']
            })
            
            # Create extraction job
            job = ExtractionJobRepository.create(session, {
                'playlist_id': playlist_id,
                'job_type': 'playlist',
                'status': 'running'
            })
            job_id = job.id
        
        # Get all videos from playlist
        console.print("[cyan]Fetching playlist videos...[/cyan]")
        playlist_videos = self.youtube_client.get_playlist_videos(playlist_id, max_videos)
        console.print(f"[green][OK][/green] Found {len(playlist_videos)} videos\n")
        
        # Filter out existing videos if requested
        videos_to_process = []
        
        if skip_existing:
            with self.db_manager.get_session() as session:
                for video in playlist_videos:
                    existing = VideoRepository.get_by_video_id(session, video['video_id'])
                    if not existing:
                        videos_to_process.append(video)
                    else:
                        # Still add to playlist items
                        PlaylistItemRepository.add_video_to_playlist(
                            session,
                            playlist_id,
                            video['video_id'],
                            video['position'],
                            datetime.fromisoformat(video['added_at'].replace('Z', '+00:00'))
                        )
            console.print(f"[yellow]Skipping {len(playlist_videos) - len(videos_to_process)} existing videos[/yellow]\n")
        else:
            videos_to_process = playlist_videos
        
        # Process videos with progress bar
        results = {
            'total': len(playlist_videos),
            'processed': 0,
            'new': 0,
            'skipped': len(playlist_videos) - len(videos_to_process),
            'failed': 0
        }
        
        # Process videos with visual counter
        for idx, video in enumerate(videos_to_process, 1):
            video_id = video['video_id']
            video_title = video.get('title', video_id)[:50]  # Truncate long titles
            
            # Visual progress header with your little dude popping up
            console.print(f"\n[bold cyan]{'='*35}[/bold cyan] [bold yellow]\\-!-/[/bold yellow] [bold cyan]{'='*35}[/bold cyan]")
            console.print(f"[bold white]Video {idx}/{len(videos_to_process)}[/bold white] | [bold blue]{playlist_title[:40]}[/bold blue]")
            console.print(f"[dim]{video_title}[/dim]")
            console.print(f"[bold cyan]{'='*80}[/bold cyan]")
            
            try:
                self.extract_single_video(video_id)
                
                # Add to playlist items
                with self.db_manager.get_session() as session:
                    PlaylistItemRepository.add_video_to_playlist(
                        session,
                        playlist_id,
                        video_id,
                        video['position'],
                        datetime.fromisoformat(video['added_at'].replace('Z', '+00:00'))
                    )
                
                results['processed'] += 1
                results['new'] += 1
            except Exception as e:
                # Sanitize error message
                error_msg = str(e).encode('ascii', 'replace').decode('ascii')
                console.print(f"[red]Error processing {video_id}: {error_msg}[/red]")
                results['failed'] += 1
        
        # Update job status
        with self.db_manager.get_session() as session:
            ExtractionJobRepository.update_status(
                session,
                job_id,
                'completed',
                videos_found=results['total'],
                videos_processed=results['processed'],
                new_videos=results['new']
            )
        
        console.print(f"\n[bold cyan]{'='*80}[/bold cyan]")
        console.print(f"[bold green]Playlist extraction complete![/bold green]")
        console.print(f"  Processed: {results['processed']}/{results['total']}")
        console.print(f"  New: {results['new']}")
        console.print(f"  Skipped: {results['skipped']}")
        if results['failed'] > 0:
            console.print(f"  [red]Failed: {results['failed']}[/red]")
        console.print(f"[bold cyan]{'='*80}[/bold cyan]")
        
        return results

File: src-ts/extractors/VideoExtractor.ts
/**
 * Main video extraction pipeline orchestrator
 * Coordinates all extractors, parsers, and database operations
 */

import { DatabaseManager } from '../database/connection.js';
import { YouTubeClient } from '../api/YouTubeClient.js';
import { TranscriptExtractor } from './TranscriptExtractor.js';
import { WhisperExtractor } from './WhisperExtractor.js';
import { DescriptionParser } from '../parsers/DescriptionParser.js';
import { GeminiParser } from '../parsers/GeminiParser.js';
import {
  VideoRepository,
  PlaylistRepository,
  PlaylistItemRepository,
  TranscriptRepository,
  EntityRepository,
} from '../database/repositories.js';
import logger from '../utils/logger.js';
import { ValidationError, AppError, DatabaseError } from '../errors/index.js';
import type { TranscriptData } from './TranscriptExtractor.js';
import type { ParsedTranscript } from '../parsers/GeminiParser.js';

/**
 * Video extraction result
 */
export interface VideoExtractionResult {
  videoData: any;
  transcriptData: TranscriptData | null;
  parsedEntities: ParsedTranscript | null;
}

/**
 * Playlist extraction result
 */
export interface PlaylistExtractionResult {
  total: number;
  processed: number;
  new: number;
  skipped: number;
  failed: number;
}

/**
 * Progress callback for playlist extraction
 */
export interface PlaylistProgressCallback {
  (progress: {
    current: number;
    total: number;
    videoId: string;
    videoTitle: string;
    status: 'processing' | 'complete' | 'failed' | 'skipped';
  }): void;
}

/**
 * Whisper progress callback
 */
export interface WhisperProgressCallback {
  (progress: {
    stage: 'downloading' | 'transcribing' | 'complete';
    percentage?: number;
    message?: string;
  }): void;
}

/**
 * VideoExtractor configuration
 */
export interface VideoExtractorConfig {
  autoTranscript?: boolean;
  autoLlmParse?: boolean;
  geminiApiKey?: string;
  languages?: string[];
  transcriptRateLimit?: number;
  enableWhisper?: boolean;
  onProgress?: PlaylistProgressCallback;
  onWhisperProgress?: WhisperProgressCallback;
}

/**
 * Main extraction pipeline for videos
 */
export class VideoExtractor {
  private readonly youtubeClient: YouTubeClient;
  private readonly db: DatabaseManager;
  private readonly autoTranscript: boolean;
  private readonly autoLlmParse: boolean;
  private readonly transcriptExtractor: TranscriptExtractor;
  private readonly descriptionParser: DescriptionParser;
  private readonly llmParser: GeminiParser | null;
  private readonly videoRepository: VideoRepository;
  private readonly transcriptRepository: TranscriptRepository;
  private readonly entityRepository: EntityRepository;
  private readonly playlistItemRepository: PlaylistItemRepository;
  private readonly playlistRepository: PlaylistRepository;
  private readonly onProgress?: PlaylistProgressCallback;

  /**
   * Create a new VideoExtractor instance
   *
   * @param youtubeClient - Authenticated YouTube API client
   * @param db - Database manager instance
   * @param config - Extractor configuration
   */
  constructor(
    youtubeClient: YouTubeClient,
    db: DatabaseManager,
    config: VideoExtractorConfig = {}
  ) {
    this.youtubeClient = youtubeClient;
    this.db = db;
    this.autoTranscript = config.autoTranscript ?? true;
    this.autoLlmParse = config.autoLlmParse ?? true;
    this.onProgress = config.onProgress;

    // Initialize repositories
    this.videoRepository = new VideoRepository(db);
    this.transcriptRepository = new TranscriptRepository(db);
    this.entityRepository = new EntityRepository(db);
    this.playlistRepository = new PlaylistRepository(db);
    this.playlistItemRepository = new PlaylistItemRepository(db);

    // Initialize Whisper if enabled
    let whisperExtractor: WhisperExtractor | undefined;
    if (config.enableWhisper) {
      whisperExtractor = new WhisperExtractor({
        enabled: true,
        onProgress: config.onWhisperProgress,
      });
      if (!whisperExtractor.isAvailable()) {
        logger.warn(
          {
            reason: whisperExtractor.getUnavailableReason(),
          },
          'Whisper enabled but not available'
        );
      }
    }

    // Initialize transcript extractor
    this.transcriptExtractor = new TranscriptExtractor({
      languages: config.languages,
      rateLimitDelay: config.transcriptRateLimit,
      whisperExtractor,
    });

    // Initialize description parser
    this.descriptionParser = new DescriptionParser();

    // Initialize LLM parser
    if (this.autoLlmParse) {
      try {
        this.llmParser = new GeminiParser(config.geminiApiKey);
      } catch (error) {
        logger.warn(
          {
            error: error instanceof Error ? error.message : String(error),
          },
          'Failed to initialize GeminiParser'
        );
        logger.warn('LLM parsing will be disabled');
        this.llmParser = null;
      }
    } else {
      this.llmParser = null;
    }

    logger.info(
      {
        autoTranscript: this.autoTranscript,
        autoLlmParse: this.autoLlmParse && this.llmParser !== null,
        hasWhisper: !!whisperExtractor,
      },
      'VideoExtractor initialized'
    );
  }

  /**
   * Extract complete data for a single video
   *
   * @param videoId - YouTube video ID
   * @param skipTranscript - Skip transcript extraction
   * @param skipLlm - Skip LLM parsing
   * @returns Extraction result or null if failed
   * @throws {ValidationError} If video ID is invalid
   */
  async extractSingleVideo(
    videoId: string,
    skipTranscript = false,
    skipLlm = false
  ): Promise<VideoExtractionResult | null> {
    // Validate video ID
    if (typeof videoId !== 'string' || !videoId.trim()) {
      throw new ValidationError('videoId must be a non-empty string');
    }

    if (!/^[a-zA-Z0-9_-]{11}$/.test(videoId)) {
      throw new ValidationError('Invalid YouTube video ID format');
    }

    logger.info({ videoId }, 'Starting video extraction');

    try {
      // Step 1: Get video metadata
      logger.debug({ videoId }, 'Fetching metadata');
      const videoData = await this.youtubeClient.getVideoDetails(videoId);

      if (!videoData) {
        logger.error({ videoId }, 'Failed to fetch video metadata');
        return null;
      }

      logger.info(
        {
          videoId,
          title: videoData.title?.substring(0, 60) || videoData.title || 'No title',
          channelTitle:
            videoData.channelTitle?.substring(0, 40) || videoData.channelTitle || 'No channel',
          duration: videoData.duration,
        },
        'Fetched video metadata'
      );

      // Step 2: Save video to database
      await this.saveVideoToDatabase(videoData);

      // Step 3: Parse description for GitHub repos and URLs
      const descriptionParsed = await this.parseDescription(
        videoId,
        videoData.title,
        videoData.description
      );

      // Step 4: Extract transcript
      let transcriptData: TranscriptData | null = null;
      if (this.autoTranscript && !skipTranscript) {
        transcriptData = await this.extractTranscript(videoId);
      }

      // Step 5: Parse with LLM
      let parsedEntities: ParsedTranscript | null = null;
      if (this.autoLlmParse && !skipLlm && transcriptData && this.llmParser) {
        parsedEntities = await this.parseTranscriptWithLlm(
          videoId,
          transcriptData,
          videoData.title
        );
      }

      logger.info({ videoId }, 'Video extraction complete');

      return {
        videoData,
        transcriptData,
        parsedEntities,
      };
    } catch (error) {
      logger.error(
        {
          videoId,
          error: error instanceof Error ? error.message : String(error),
          stack: error instanceof Error ? error.stack : undefined,
        },
        'Failed to extract video'
      );

      throw new AppError('Video extraction failed', {
        cause: error,
        code: 'EXTRACTION_FAILED',
      });
    }
  }

  /**
   * Extract all videos from a playlist
   *
   * @param playlistId - YouTube playlist ID
   * @param skipExisting - Skip videos already in database
   * @param maxVideos - Maximum number of videos to process
   * @returns Extraction summary
   * @throws {ValidationError} If playlist ID is invalid
   */
  async extractPlaylist(
    playlistId: string,
    skipExisting = true,
    maxVideos?: number
  ): Promise<PlaylistExtractionResult> {
    // Validate playlist ID
    if (typeof playlistId !== 'string' || !playlistId.trim()) {
      throw new ValidationError('playlistId must be a non-empty string');
    }

    logger.info({ playlistId, skipExisting, maxVideos }, 'Starting playlist extraction');

    try {
      // Get playlist info
      const playlistResponse = await this.youtubeClient.getPlaylists();
      const playlist = playlistResponse.playlists.find((p) => p.id === playlistId);

      if (!playlist) {
        throw new ValidationError('Playlist not found');
      }

      logger.info(
        {
          playlistId,
          title: playlist.title.substring(0, 60),
          videoCount: playlist.itemCount,
        },
        'Found playlist'
      );

      // Save playlist to database
      await this.savePlaylistToDatabase(playlist);

      // Get all videos from playlist — paginate until YouTube returns no more.
      // YouTube's playlistItems.list maxResults caps at 50 per page, so a playlist
      // with >50 videos needs multiple calls. Without pagination we silently dropped
      // tail videos (observed on the meTube Ai playlist: 86 in YouTube, 50 received).
      logger.debug({ playlistId, maxVideos }, 'Fetching playlist videos (paginating)');
      const PAGE_SIZE = 50;
      const allItems: typeof playlist.itemCount extends never
        ? never[]
        : Awaited<ReturnType<typeof this.youtubeClient.getPlaylistVideos>>['items'] = [];
      let pageToken: string | undefined = undefined;
      let pages = 0;
      do {
        const remaining = maxVideos !== undefined ? maxVideos - allItems.length : PAGE_SIZE;
        if (maxVideos !== undefined && remaining <= 0) break;
        const pageSize = Math.min(PAGE_SIZE, remaining > 0 ? remaining : PAGE_SIZE);
        const pageResponse = await this.youtubeClient.getPlaylistVideos(
          playlistId,
          pageSize,
          pageToken
        );
        allItems.push(...pageResponse.items);
        pageToken = pageResponse.nextPageToken;
        pages += 1;
      } while (pageToken);
      const playlistVideos = maxVideos !== undefined ? allItems.slice(0, maxVideos) : allItems;
      logger.info(
        {
          playlistId,
          pages,
          fetched: allItems.length,
          considered: playlistVideos.length,
        },
        'Fetched playlist videos'
      );

      // Filter out existing videos if requested
      let videosToProcess = playlistVideos;
      let skippedCount = 0;

      if (skipExisting) {
        const existingVideos = new Set<string>();

        for (const video of playlistVideos) {
          const existing = this.videoRepository.getByVideoId(video.videoId);
          const hasTranscript = existing
            ? this.transcriptRepository.getByVideoId(video.videoId)
            : null;

          // Only skip if BOTH video AND transcript exist
          if (existing && hasTranscript) {
            existingVideos.add(video.videoId);
            // Still save to playlist items
            await this.savePlaylistItem(playlistId, video);
          }
        }

        videosToProcess = playlistVideos.filter((v) => !existingVideos.has(v.videoId));
        skippedCount = playlistVideos.length - videosToProcess.length;

        logger.info(
          {
            total: playlistVideos.length,
            toProcess: videosToProcess.length,
            skipped: skippedCount,
          },
          'Filtered existing videos (with transcripts)'
        );
      }

      // Process videos
      const results: PlaylistExtractionResult = {
        total: playlistVideos.length,
        processed: 0,
        new: 0,
        skipped: skippedCount,
        failed: 0,
      };

      for (let idx = 0; idx < videosToProcess.length; idx++) {
        const video = videosToProcess[idx];
        const videoId = video.videoId;

        logger.info(
          {
            index: idx + 1,
            total: videosToProcess.length,
            videoId,
            title: video.title?.substring(0, 50),
          },
          'Processing video'
        );

        // Report progress - processing
        if (this.onProgress) {
          this.onProgress({
            current: idx + 1,
            total: videosToProcess.length,
            videoId,
            videoTitle: video.title || 'Unknown',
            status: 'processing',
          });
        }

        try {
          await this.extractSingleVideo(videoId);

          // Add to playlist items
          await this.savePlaylistItem(playlistId, video);

          results.processed++;
          results.new++;

          // Report progress - complete
          if (this.onProgress) {
            this.onProgress({
              current: idx + 1,
              total: videosToProcess.length,
              videoId,
              videoTitle: video.title || 'Unknown',
              status: 'complete',
            });
          }
        } catch (error) {
          logger.error(
            {
              videoId,
              error: error instanceof Error ? error.message : String(error),
            },
            'Failed to process video'
          );
          results.failed++;

          // Report progress - failed
          if (this.onProgress) {
            this.onProgress({
              current: idx + 1,
              total: videosToProcess.length,
              videoId,
              videoTitle: video.title || 'Unknown',
              status: 'failed',
            });
          }
        }
      }

      logger.info(results, 'Playlist extraction complete');

      return results;
    } catch (error) {
      logger.error(
        {
          playlistId,
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to extract playlist'
      );

      throw new AppError('Playlist extraction failed', {
        cause: error,
        code: 'PLAYLIST_EXTRACTION_FAILED',
      });
    }
  }

  /**
   * Save video to database
   * Saves video metadata to videos table and statistics to video_statistics table
   */
  private async saveVideoToDatabase(videoData: any): Promise<void> {
    try {
      const videoId = videoData.id || videoData.video_id;

      // Map YouTubeVideo (camelCase) to database schema (snake_case)
      const publishedAt = videoData.publishedAt || videoData.published_at;
      const videoRecord = {
        video_id: videoId,
        title: videoData.title,
        description: videoData.description || '',
        channel_id: videoData.channelId || videoData.channel_id,
        channel_title: videoData.channelTitle || videoData.channel_title,
        published_at: publishedAt ? new Date(publishedAt).toISOString() : new Date().toISOString(),
        duration: videoData.duration || 'PT0S',
        duration_seconds: this.parseDurationToSeconds(videoData.duration) || 0,
        is_short: Boolean(videoData.is_short),
        category_id: videoData.category_id || null,
        definition: videoData.definition || 'sd',
        caption: Boolean(videoData.caption),
        licensed_content: Boolean(videoData.licensed_content),
      };

      // Save video metadata
      this.videoRepository.createOrUpdate(videoRecord);

      // Save video statistics separately
      const viewCount = videoData.viewCount || videoData.view_count || 0;
      const likeCount = videoData.likeCount || videoData.like_count || 0;
      const commentCount = videoData.commentCount || videoData.comment_count || 0;

      if (viewCount > 0 || likeCount > 0 || commentCount > 0) {
        const statisticRecord = {
          video_id: videoId,
          view_count: viewCount,
          like_count: likeCount,
          comment_count: commentCount,
          recorded_at: new Date(),
        };

        // Use raw SQL for now (can create StatisticsRepository later if needed)
        this.db.run(
          `INSERT INTO video_statistics (video_id, view_count, like_count, comment_count, recorded_at)
           VALUES (?, ?, ?, ?, ?)`,
          [
            statisticRecord.video_id,
            statisticRecord.view_count,
            statisticRecord.like_count,
            statisticRecord.comment_count,
            statisticRecord.recorded_at.toISOString(),
          ]
        );
      }

      logger.debug(
        {
          video_id: videoId,
        },
        'Saved video to database'
      );
    } catch (error) {
      const errorDetails =
        error instanceof Error
          ? { message: error.message, stack: error.stack, cause: (error as any).cause }
          : String(error);

      logger.error(
        {
          video_id: videoData.id || videoData.video_id,
          error: errorDetails,
        },
        'Failed to save video to database'
      );

      console.error('\n=== FULL DATABASE ERROR ===');
      console.error('Video ID:', videoData.id || videoData.video_id);
      console.error('Error:', error);
      if (error instanceof Error) {
        console.error('Stack:', error.stack);
        console.error('Cause:', (error as any).cause);
      }
      console.error('=== END ERROR ===\n');

      throw new DatabaseError('Failed to save video', { cause: error });
    }
  }

  /**
   * Parse ISO 8601 duration to seconds
   */
  private parseDurationToSeconds(duration: string): number {
    if (!duration || !duration.startsWith('PT')) {
      return 0;
    }

    const match = duration.match(/PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?/);
    if (!match) {
      return 0;
    }

    const hours = parseInt(match[1] || '0', 10);
    const minutes = parseInt(match[2] || '0', 10);
    const seconds = parseInt(match[3] || '0', 10);

    return hours * 3600 + minutes * 60 + seconds;
  }

  /**
   * Parse description for entities
   */
  private async parseDescription(
    videoId: string,
    title: string,
    description: string
  ): Promise<void> {
    try {
      logger.debug({ videoId }, 'Parsing description');

      const descriptionParsed = this.descriptionParser.parse(title, description);

      if (descriptionParsed.github_repos.length > 0 || descriptionParsed.websites.length > 0) {
        logger.info(
          {
            videoId,
            github_repos: descriptionParsed.github_repos.length,
            websites: descriptionParsed.websites.length,
          },
          'Found entities in description'
        );

        // Add entities from description
        const descEntities = this.descriptionParser.extractEntitiesForDatabase(descriptionParsed);
        if (descEntities.length > 0) {
          // Convert null to undefined for url field
          const entities = descEntities.map((e) => ({
            ...e,
            url: e.url || undefined,
          }));
          this.entityRepository.addEntities(videoId, entities);
        }
      } else {
        logger.debug({ videoId }, 'No links found in description');
      }
    } catch (error) {
      logger.error(
        {
          videoId,
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to parse description'
      );
      // Don't throw - continue with extraction
    }
  }

  /**
   * Extract transcript for video
   */
  private async extractTranscript(videoId: string): Promise<TranscriptData | null> {
    try {
      logger.debug({ videoId }, 'Extracting transcript');

      const transcriptData = await this.transcriptExtractor.extract(videoId);

      if (transcriptData) {
        const source = transcriptData.from_whisper ? 'Whisper' : 'YouTube';
        logger.info(
          {
            videoId,
            source,
            length: transcriptData.full_text.length,
            segments: transcriptData.segments.length,
          },
          'Transcript extracted'
        );

        // Save transcript to database
        this.transcriptRepository.create(videoId, transcriptData);

        return transcriptData;
      } else {
        logger.warn({ videoId }, 'No transcript available');
        return null;
      }
    } catch (error) {
      logger.error(
        {
          videoId,
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to extract transcript'
      );
      return null;
    }
  }

  /**
   * Parse transcript with LLM
   */
  private async parseTranscriptWithLlm(
    videoId: string,
    transcriptData: TranscriptData,
    videoTitle: string
  ): Promise<ParsedTranscript | null> {
    if (!this.llmParser) {
      return null;
    }

    try {
      logger.debug({ videoId }, 'Parsing with Gemini AI');

      const parsedEntities = await this.llmParser.parseTranscript(
        transcriptData.full_text,
        videoTitle
      );

      logger.info(
        {
          videoId,
          topics: parsedEntities.topics.length,
          github_repos: parsedEntities.github_repos.length,
          websites: parsedEntities.websites.length,
          people: parsedEntities.people.length,
        },
        'Parsed transcript with Gemini'
      );

      // Delete old entities
      this.entityRepository.deleteByVideo(videoId);

      // Add new entities
      const entities = this.llmParser.extractEntitiesForDatabase(parsedEntities);
      // Convert null to undefined for url field
      const entitiesFixed = entities.map((e) => ({
        ...e,
        url: e.url || undefined,
      }));
      this.entityRepository.addEntities(videoId, entitiesFixed);

      return parsedEntities;
    } catch (error) {
      logger.error(
        {
          videoId,
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to parse transcript with LLM'
      );
      return null;
    }
  }

  /**
   * Save playlist to database
   */
  private async savePlaylistToDatabase(playlist: any): Promise<void> {
    try {
      const playlistRecord = {
        playlist_id: playlist.id,
        title: playlist.title,
        description: playlist.description || '',
        video_count: playlist.video_count || 0,
        enabled: true,
      };

      this.playlistRepository.createOrUpdate(playlistRecord);

      logger.debug(
        {
          playlist_id: playlist.id,
        },
        'Saved playlist to database'
      );
    } catch (error) {
      logger.error(
        {
          playlist_id: playlist.id,
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to save playlist to database'
      );

      throw new DatabaseError('Failed to save playlist', { cause: error });
    }
  }

  /**
   * Save playlist item to database
   */
  private async savePlaylistItem(
    playlistId: string,
    video: { videoId: string; position?: number }
  ): Promise<void> {
    try {
      this.playlistItemRepository.addVideoToPlaylist(playlistId, video.videoId, video.position);
      logger.debug({ playlistId, videoId: video.videoId }, 'Saved playlist_items linkage');
    } catch (error) {
      logger.error(
        {
          playlistId,
          videoId: video.videoId,
          error: error instanceof Error ? error.message : String(error),
        },
        'Failed to save playlist_items linkage'
      );
      // Don't throw - a playlist-link failure shouldn't kill the whole extraction.
    }
  }
}
